README.md
# thingtalk-units

[![Build Status](https://travis-ci.com/stanford-oval/thingtalk-units.svg?branch=master)](https://travis-ci.org/stanford-oval/thingtalk-units) [![Coverage Status](https://coveralls.io/repos/github/stanford-oval/thingtalk-units/badge.svg?branch=master)](https://coveralls.io/github/stanford-oval/thingtalk-units?branch=master) [![Dependency Status](https://david-dm.org/stanford-oval/thingtalk-units/status.svg)](https://david-dm.org/stanford-oval/thingtalk-units) [![Greenkeeper badge](https://badges.greenkeeper.io/stanford-oval/thingtalk-units.svg)](https://greenkeeper.io/) [![Language grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/stanford-oval/thingtalk-units.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/stanford-oval/thingtalk-units/context:javascript)

A small library to convert and normalize between measurement units. This library
is used by the ThingTalk language to support measurements.

The library converts to and from _base units_ of various supported types, and various
additional units for each base unit.

## Usage

```javascript
const TTUnits = require('thingtalk-units');
```

## API

### function normalizeUnit(unit : String) : String

Convert a unit name to the base unit, which is a standard unit for the same property.
This method throws if the unit is invalid.
Example:

```javascript
TTUnits.normalizeUnit('F') === 'C'
TTUnits.normalizeUnit('km') === 'm'
```

### function transformToBaseUnit(value : Number, unit : String) : Number

Convert the value `value`, expressed in the unit `unit`, in the same amount in the corresponding
base unit.
Example:

```javascript
TTUnits.transformToBaseUnit(1, 'km') === 1000
```

### function transformFromBaseUnit(value : Number, unit : String) : Number

Convert the value `value`, expressed in the base unit, to the unit `unit`.
This is the inverse operation to `transformToBaseUnit()`.
Example:

```javascript
TTUnits.transformToBaseUnit(1000, 'km') === 1
```

### const BaseUnit : Array<string>

The list of the supported base units.

## Supported Units

### Base Units

* `ms`: time (milliseconds)
* `m`: distance (meters)
* `mps`: speed (meters per second)
* `kg`: mass (kilograms)
* `Pa`: pressure (Pascal)
* `C`: temperature (Celsius)
* `kcal`: energy (kilocalories)
* `byte`: data size (Byte)
* `W`: power (Watt)
* `lm`: luminous flux (lumen)
* `lx`: illuminance (lux)

The choice of base units is mostly historical. Some units follow SI convention (mps, kg, Pa),
some follow common usage (kcal, C).

### Additional Units

| Unit                              | Base Unit | Physical quantity |
| --------------------------------- | --------- | ----------------- |
| `s` (seconds)                     | `ms`      | time              |
| `min` (minutes)                   | `ms`      | time              |
| `h` (hours)                       | `ms`      | time              |
| `day` (days)                      | `ms`      | time              |
| `week` (weeks)                    | `ms`      | time              |
| `mon` (business month = 30 days)  | `ms`      | time              |
| `year` (business year = 365 days) | `ms`      | time              |
| `km` (kilometers)                 | `m`       | distance          |
| `mm` (millimeters)                | `m`       | distance          |
| `cm` (centimeters)                | `m`       | distance          |
| `mi` (miles)                      | `m`       | distance          |
| `in` (inches)                     | `m`       | distance          |
| `kmph` (kilometers per hour)      | `mps`     | speed             |
| `mph` (miles per hour)            | `mps`     | speed             |
| `g` (grams)                       | `kg`      | mass              |
| `lb` (US pound)                   | `kg`      | mass              |
| `oz` (US ounce)                   | `kg`      | mass              |
| `bar` (bar)                       | `Pa`      | pressure          |
| `psi` (pounds per square inch)    | `Pa`      | pressure          |
| `mmHg` (millimeters of mercury)   | `Pa`      | pressure          |
| `inHg` (inches of mercury)        | `Pa`      | pressure          |
| `F` (Fahrenheit)                  | `C`       | temperature       |
| `K` (Kelvin)                      | `C`       | temperature       |
| `kJ` (kilojoule)                  | `kcal`    | energy            |
| `KB` (kilobyte)                   | `byte`    | data size         |
| `KiB` (kibibyte)                  | `byte`    | data size         |
| `MB` (megabyte)                   | `byte`    | data size         |
| `MiB` (mebibyte)                  | `byte`    | data size         |
| `GB` (gigabyte)                   | `byte`    | data size         |
| `GiB` (gibibyte)                  | `byte`    | data size         |
| `TB` (terabyte)                   | `byte`    | data size         |
| `TiB` (tebibyte)                  | `byte`    | data size         |
| `kW` (kilowatt)                   | `W`       | power             |


HISTORY
0.2.0
=====

* Added units for surface area and volume [#17].
* Updated dependencies [#1, #5, #6, #7, #8, #10, #11, #12, #13, #14,
  #15, #16]. 

0.1.0
=====

Initial release.


test.js
// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// Copyright 2015-2019 The Board of Trustees of the Leland Stanford Junior University
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
//
// See LICENSE for details
"use strict";

// Unit Tests! Cause I'm testing units! Get it? Get it?

const assert = require('assert');
const TTUnits = require('.');

function test(v1, u1, v2, u2) {
    assert.strictEqual(TTUnits.normalizeUnit(u1),
        TTUnits.normalizeUnit(u2));

    let vn1 = TTUnits.transformToBaseUnit(v1, u1);
    let vn2 = TTUnits.transformToBaseUnit(v2, u2);
    assert(Math.abs(vn1 - vn2) < 0.00001);

    let v1recovered = TTUnits.transformFromBaseUnit(vn1, u1);
    assert(Math.abs(v1 - v1recovered) < 0.00001);

    let v2recovered = TTUnits.transformFromBaseUnit(vn2, u2);
    assert(Math.abs(v2 - v2recovered) < 0.00001);
}

function main() {
    test(1, 'KB', 1000, 'byte');
    test(1000, 'KB', 1000000, 'byte');
    test(1, 'kcal', 4.183995, 'kJ');
    test(100, 'C', 212, 'F');
    test(32, 'F', 0, 'C');
    test(100, 'C', 373.15, 'K');
    test(1, 'day', 86400, 's');
    test(1, 'week', 7, 'day');
    test(1, 'mon', 30, 'day');
    test(1, 'kg', 2.20462, 'lb');
    test(1, 'ft', 30.48, 'cm');
    test(1, 'ft2', 929.0304, 'cm2');
    test(1, 'ft3', 28316.846592, 'cm3');
    test(1, 'kW', 1000, 'W');
    test(1, 'W', 0.001, 'kW');

    assert.strictEqual(TTUnits.normalizeUnit(''), '');
    assert.throws(() => TTUnits.normalizeUnit('bad'));
}
main();


index.js
// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// Copyright 2015-2019 The Board of Trustees of the Leland Stanford Junior University
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
//
// See LICENSE for details
"use strict";

exports.BaseUnits = ['ms', 'm', 'm2', 'm3', 'mps', 'kg', 'Pa', 'C', 'kcal', 'byte', 'W', 'lm', 'lx', 'dB', 'dBm'];

const UnitsToBaseUnit = {
    // time
    'ms': 'ms', // base unit for time is milliseconds, because +new Date gives milliseconds
    's': 'ms',
    'min': 'ms',
    'h': 'ms',
    'day': 'ms',
    'week': 'ms',
    'mon': 'ms', // business month, aka exactly 30 days
    'year': 'ms', // business year (365 days exactly, no leap years)
    // length
    'm': 'm',
    'km': 'm',
    'mm': 'm',
    'cm': 'm',
    'mi': 'm',
    'in': 'm',
    'ft': 'm',
    // area
    'm2': 'm2',
    'km2': 'm2',
    'mm2': 'm2',
    'cm2': 'm2',
    'mi2': 'm2',
    'in2': 'm2',
    'ft2': 'm2',
    // volume
    'm3': 'm3',
    'km3': 'm3',
    'mm3': 'm3',
    'cm3': 'm3',
    'mi3': 'm3',
    'in3': 'm3',
    'ft3': 'm3',
    'gal': 'm3',
    'galuk': 'm3',
    'qt': 'm3',
    'qtuk': 'm3',
    'pint': 'm3',
    'pintuk': 'm3',
    'l': 'm3',
    'hl': 'm3',
    'cl': 'm3',
    'ml': 'm3',
    'tsp': 'm3',
    'tbsp': 'm3',
    'cup': 'm3',
    'floz': 'm3',
    // speed
    'mps': 'mps', // meters per second, usually written as m/s but m/s is not an identifier
    'kmph': 'mps',
    'mph': 'mps',
    // weight
    'kg': 'kg',
    'g': 'kg',
    'mg': 'kg',
    'lb': 'kg',
    'oz': 'kg',
    // pressure (for weather or blood)
    'Pa': 'Pa',
    'bar': 'Pa',
    'psi': 'Pa',
    'mmHg': 'Pa',
    'inHg': 'Pa',
    'atm': 'Pa',
    // temperature
    'C': 'C',
    'F': 'C',
    'K': 'C',
    // energy
    'kcal': 'kcal',
    'kJ': 'kcal',
    // file and memory sizes
    'byte': 'byte',
    'KB': 'byte',
    'KiB': 'byte',
    'MB': 'byte',
    'MiB': 'byte',
    'GB': 'byte',
    'GiB': 'byte',
    'TB': 'byte',
    'TiB': 'byte',
    // power
    'W': 'W',
    'kW': 'W',
    // luminous flux, luminous power
    'lm': 'lm',
    // luminous emittance (luminous flux emitted from a surface)
    'lx': 'lx',
    // decibel
    'dB': 'dB',
    // decibel-milliwatts
    'dBm': 'dBm'
};

const UnitsTransformToBaseUnit = {
    'ms': 1,
    's': 1000,
    'min': 60 * 1000,
    'h': 3600 * 1000,
    'day': 86400 * 1000,
    'week': 86400 * 7 * 1000,
    'mon': 86400 * 30 * 1000,
    'year': 86400 * 365 * 1000,
    'm': 1,
    'km': 1000,
    'mm': 1/1000,
    'cm': 1/100,
    'mi': 1609.344,
    'in': 0.0254,
    'ft': 0.3048,
    'm2': 1,
    'km2': 1000 * 1000,
    'mm2': 1/1000/1000,
    'cm2': 1/100/100,
    'mi2': 1609.344 * 1609.344,
    'in2': 0.0254 * 0.0254,
    'ft2': 0.3048 * 0.3048,
    'm3': 1,
    'km3': 1000 * 1000 * 1000,
    'mm3': 1/1000/1000/1000,
    'cm3': 1/100/100/100,
    'mi3': 1609.344 * 1609.344 * 1609.344,
    'in3': 0.0254 * 0.0254 * 0.0254,
    'ft3': 0.3048 * 0.3048 * 0.3048,
    'gal': 0.00378541,
    'galuk': 0.00454609,
    'qt': 0.000946353,
    'qtuk': 0.00113652,
    'pint': 0.000473176,
    'pintuk': 0.000568261,
    'l': 0.001,
    'hl': 0.1,
    'cl': 0.00001,
    'ml': 0.000001,
    'tsp': 0.00000492892,
    'tbsp': 0.0000147868,
    'cup': 0.000236588,
    'floz': 0.0000295735,
    'mps': 1,
    'kmph': 0.27777778,
    'mph': 0.44704,
    'kg': 1,
    'g': 1/1000,
    'mg': 1e-6,
    'lb': 0.45359237,
    'oz': 0.028349523,
    'Pa': 1,
    'bar': 100000,
    'psi': 6894.7573,
    'mmHg': 133.32239,
    'inHg': 3386.3886,
    'atm': 101325,
    'C': 1,
    'F': function(x) { return (x - 32)/1.8; },
    'K': function(x) { return x - 273.15; },
    'kcal': 1,
    'kJ': 0.239006,
    'byte': 1,
    'KB': 1000,
    'KiB': 1024,
    'MB': 1000*1000,
    'MiB': 1024*1024,
    'GB': 1000*1000*1000,
    'GiB': 1024*1024*1024,
    'TB': 1000*1000*1000*1000,
    'TiB': 1024*1024*1024*1024,
    'W': 1,
    'kW': 1000,
    'lm': 1,
    'lx': 1,
    'dB': 1,
    'dBm': 1
};

const UnitsInverseTransformFromBaseUnit = {
    'F': function(x) { return x*1.8 + 32; },
    'K': function(x) { return x + 273.15; }
};

exports.transformToBaseUnit = function(value, unit) {
    var transform = UnitsTransformToBaseUnit[unit];
    if (typeof transform === 'function')
        return transform(value);
    else
        return value * transform;
};

exports.transformFromBaseUnit = function(value, unit) {
    var coeff = UnitsTransformToBaseUnit[unit];
    if (typeof coeff === 'function')
        return UnitsInverseTransformFromBaseUnit[unit](value);
    else
        return ((1/coeff)*value);
};

exports.normalizeUnit = function(unit) {
    if (unit === '')
        return '';
    var baseunit = UnitsToBaseUnit[unit];
    if (baseunit === undefined)
        throw new TypeError('Invalid unit ' + unit);
    return baseunit;
};
